arraycontext/impl/pytato.py

```diff
@@ -0,0 +1,477 @@
+
+
+class PytatoArrayContext(ArrayContext):
+    """
+    A :class:`ArrayContext` that uses :mod:`pytato` data types to represent
+    the DOF arrays targeting OpenCL for offloading operations.
+
+    .. attribute:: queue
+
+        A :class:`pyopencl.CommandQueue`.
+
+    .. attribute:: allocator
+
+        A :mod:`pyopencl` memory allocator. Can also be None (default) or False
+        to use the default allocator.
+    """
+
+    def __init__(self, queue, allocator=None):
+        super().__init__()
+        self.queue = queue
+        self.allocator = allocator
+        self.np = self._get_fake_numpy_namespace()
+
+    def _get_fake_numpy_namespace(self):
+        return _PytatoFakeNumpyNamespace(self)
+
+    # {{{ ArrayContext interface
+
+    def clone(self):
+        return type(self)(self.queue, self.allocator)
+
+    def empty(self, shape, dtype):
+        raise ValueError("PytatoArrayContext does not support empty")
+
+    def zeros(self, shape, dtype):
+        import pytato as pt
+        return pt.zeros(shape, dtype)
+
+    def from_numpy(self, np_array: np.ndarray):
+        import pytato as pt
+        import pyopencl.array as cla
+        cl_array = cla.to_device(self.queue, np_array)
+        return pt.make_data_wrapper(cl_array)
+
+    def to_numpy(self, array):
+        cl_array = self.freeze(array)
+        return cl_array.get(queue=self.queue)
+
+    def call_loopy(self, program, **kwargs):
+        from pytato.loopy import call_loopy  # type: ignore
+        import pyopencl.array as cla
+        entrypoint, = set(program.callables_table)
+
+        # thaw frozen arrays
+        kwargs = {kw: (self.thaw(arg) if isinstance(arg, cla.Array) else arg)
+                  for kw, arg in kwargs.items()}
+
+        return call_loopy(program, kwargs, entrypoint)
+
+    def freeze(self, array):
+        import pytato as pt
+        import pyopencl.array as cla
+
+        if isinstance(array, pt.Placeholder):
+            raise ValueError("freezing placeholder would return garbage valued"
+                    " arrays")
+        if isinstance(array, cla.Array):
+            return array.with_queue(None)
+        if not isinstance(array, pt.Array):
+            raise TypeError("PytatoArrayContext.freeze invoked with non-pt "
+                            f"array of type '{type(array)}'")
+
+        prg = pt.generate_loopy(array, cl_device=self.queue.device)
+        evt, (cl_array,) = prg(self.queue)
+        evt.wait()
+
+        return cl_array.with_queue(None)
+
+    def thaw(self, array):
+        import pytato as pt
+        import pyopencl.array as cla
+
+        if not isinstance(array, cla.Array):
+            raise TypeError("PytatoArrayContext.thaw expects CL arrays, got "
+                    f"{type(array)}")
+
+        return pt.make_data_wrapper(array.with_queue(self.queue))
+
+    # }}}
+
+    def compile(self, f: Callable[[Any], Any],
+                inputs_like: Tuple[Union[Number, np.ndarray], ...]
```

Comment:
This `inputs_like` thing is a crutch, IMO, and I'm not convinced it's needed. Instead, we could wait until the user calls the "compiled" function and use the data provided in this role. That way, nobody has to make up "fake" data with unclear semantics.

Reply:
Thanks, agreed: a7b372a.